kernel/siw_cm.c

```diff
@@ -740,6 +819,62 @@ static int siw_proc_mpareply(struct siw_cep *cep)
 		rv = -EINVAL;
 		goto out;
 	}
+
+	cep->p2ptype = SIW_MPAV2_P2P_DISABLED;
+
+	if (__mpa_rr_revision(rep->params.bits) == MPA_REVISION_2) {
+		if (rep->params.bits & MPA_RR_FLAG_ENHANCED) {
+			cep->enhanced_rdma_connection = 1;
+		} else {
+			cep->enhanced_rdma_connection = 0;
+		}
+	
+		if (cep->enhanced_rdma_connection) {
+			v2 = (struct mpa_v2_data *)cep->mpa.pdata;
+			rep_ird = ntohs(v2->ird) & MPA_IRD_ORD_MASK;
+			rep_ord = ntohs(v2->ord) & MPA_IRD_ORD_MASK;
+
+			if (cep->ird < rep_ord) {
+				if (RELAXED_IRD_NEGOTIATION && rep_ord <=
+				    cep->sdev->attrs.max_ord)
+					cep->ird = rep_ord;
+				else
+					ird_insufficient = 1;
+			} else if (cep->ird > rep_ord) {
+				cep->ird = rep_ord;
+			}
+
+			if (cep->ord > rep_ird) {
+				if (RELAXED_IRD_NEGOTIATION)
+					cep->ord = rep_ird;
+				else
+					ird_insufficient = 1;
+			}
+
+			if (ird_insufficient) {
+				rv = -ENOMEM;
+				cep->ird = rep_ord;
+				cep->ord = rep_ird;
+			}
+
+			cep->mpa.enhanced_conn_data.ird = htons(cep->ird) | 
+						MPA_V2_PEER_TO_PEER;
+			cep->mpa.enhanced_conn_data.ord = htons(cep->ord);
+			if (v2->ird & MPA_V2_PEER_TO_PEER) {
+				if (v2->ord & MPA_V2_RDMA_WRITE_RTR) {
+					cep->p2ptype = SIW_MPAV2_RDMA_WRITE_RTR;
+					cep->mpa.enhanced_conn_data.ord |=
+						MPA_V2_RDMA_WRITE_RTR;
+				} else { if (v2->ord & MPA_V2_RDMA_READ_RTR)
```

Comment:
just saw this late. seems to be a typo. i'd expect 
} else if (v2->ord & MPA_V2_RDMA_READ_RTR) { 
                            cep->p2ptype = SIW_MPAV2_RDMA_READ_RTR;
                            cep->mpa.enhanced_conn_data.ord |= MPA_V2_RDMA_READ_RTR;
}
right?
i fix that in my rebased version